### PR TITLE
add apFirst and apSecond

### PR DESCRIPTION
**Pull request by zhangchiqing**
Hi @davidchambers ,

I just moved the changes from https://github.com/sanctuary-js/sanctuary/pull/270#issuecomment-258389389 to here.

I didn't change the function names, as they are still in discussion.

Please review. 

I will create another PR to https://github.com/sanctuary-js/sanctuary/pull/216 once this PR has been merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -50,7 +50,7 @@ This project provides:
 
 ## API
 
-<h4 name="TypeClass"><code><a href="https://github.com/sanctuary-js/sanctuary-type-classes/blob/v1.0.0/index.js#L105">TypeClass :: (String, Array TypeClass, a -> Boolean) -> TypeClass</a></code></h4>
+<h4 name="TypeClass"><code><a href="https://github.com/sanctuary-js/sanctuary-type-classes/blob/v1.0.0/index.js#L112">TypeClass :: (String, Array TypeClass, a -> Boolean) -> TypeClass</a></code></h4>
```
Comment (davidchambers) [inline README.md]: As mentioned in the [__CONTRIBUTING.md__][1], it's not necessary to update __README.md__ manually. ;)


[1]: https://github.com/sanctuary-js/sanctuary-type-classes/blob/master/CONTRIBUTING.md

File: index.js [hunk 2]
```diff
@@ -88,6 +88,13 @@
   //  identity :: a -> a
   var identity = function(x) { return x; };
 
+  // kCombinator :: a -> b -> a
+  var kCombinator = function(x) {
```
Comment (davidchambers) [inline index.js]: Let's name this `constant` for consistency with `identity`.

File: index.js [hunk 3]
```diff
@@ -1056,6 +1063,42 @@
     return ap(ap(map(f, x), y), z);
   };
 
+  //# apFirst :: Apply f => f a -> f b -> f a
+  //.
+  //. Combine two effectful actions, keeping only the result of the first.
+  //. Equivalent to Haskell's <* function.
```
Comment (davidchambers) [inline index.js]: Let's write `` `(<*)` `` rather than `<*`.

File: index.js [hunk 4]
```diff
@@ -1056,6 +1063,42 @@
...
+  //.
+  //. See also [`apSecond`](#apSecond).
+  //.
+  //. > apFirst([1, 2], [3, 4])
+  //. [1, 1, 2, 2]
+  //.
+  //. ```javascript
```
Comment (davidchambers) [inline index.js]: This needs to move up a few lines.

**Comment by zhangchiqing** [comment]
Thanks for the review @davidchambers .

I've updated the PR for all your comments. And reverted the changes to README.md.

Please let me know they are good :)

File: index.js [hunk 5]
```diff
@@ -88,6 +88,13 @@
   //  identity :: a -> a
   var identity = function(x) { return x; };
 
+  // constant :: a -> b -> a
+  var constant = function(x) {
```
Comment (davidchambers) [inline index.js]: Please add one more space between `//` and `constant` to align the function name in type line with the identifier on the following line. :)

Also, for lack of a better system the helper functions are defined in alphabetical order, so `constant` should be defined before `identity`.

File: index.js [hunk 6]
```diff
@@ -1056,6 +1063,42 @@
     return ap(ap(map(f, x), y), z);
   };
 
+  //# apFirst :: Apply f => (f a, f b) -> f a
+  //.
+  //. Combine two effectful actions, keeping only the result of the first.
+  //. Equivalent to Haskell's (<*) function.
```
Comment (davidchambers) [inline index.js]: Let's wrap the function name in backticks so it will be rendered as "`(<*)`" rather than "(<*)".

**Comment by zhangchiqing** [comment]
Good advices! 

Just updated, please take a look again, thanks @davidchambers

**Comment by davidchambers** [comment]
Thanks very much for submitting this pull request, @zhangchiqing, and for updating it quickly in response to my suggestions.

I'd like to leave this open for a day or two while we decide on the most appropriate names, but otherwise this looks ready to merge. :)

**Comment by zhangchiqing** [comment]
Good to know :)

Yes, I agree. Please let me know if we need to rename them.

Thanks again for the code review.

**Comment by safareli** [comment]
@scott-christopher I agree left/right could be confusing because of Either type, in that case I think `first/last` is  better option.

**Comment by davidchambers** [comment]
Are you happy with this pull request as it stands, @safareli?

**Comment by safareli** [comment]
I would like if we use `apLast` instead of `apSecond`, but if not it's not that big issue.

**Comment by davidchambers** [comment]
They're essentially binary operators, aren't they? I think of the blanks in `_ *> _` as _left-hand operand_ and _right-hand operand_, or as _first operand_ and _second operand_. The word _last_ suggests to me we're dealing with a variable number of things rather than exactly two things.

**Comment by safareli** [comment]
Agree!

File: index.js [hunk 7]
```diff
@@ -1272,6 +1315,8 @@
     ap: ap,
     lift2: lift2,
     lift3: lift3,
+    apFirst: apFirst,
+    apSecond: apSecond,
```
Comment (davidchambers) [inline index.js]: One last thing: I think `apFirst` and `apSecond` should come between `ap` and `lift`. Could you move the definitions, the exports, and the tests?

Comment (safareli) [inline index.js]: I think you can still have access to lift2 if you move apFirst because of js hoisting

Comment (davidchambers) [inline index.js]: Good point, @zhangchiqing. The location you chose is perfect! I'll merge this when I get home. :)

**Comment by davidchambers** [comment]
Thanks again for this nice contribution, @zhangchiqing. Don't worry about creating a Sanctuary pull request; I will update sanctuary-js/sanctuary#216 to include `S.apFirst` and `S.apSecond`.

:deciduous_tree:

**Comment by zhangchiqing** [comment]
Awesome! I'm more than happy to contribute :)

**Comment by zhangchiqing** [comment]
One more question:

Will the [README.md](https://github.com/sanctuary-js/sanctuary-type-classes/blob/master/README.md) be updated with the new functions once this PR gets merged? 

Should we run `make all` manually?

**Comment by safareli** [comment]
readme is updated when version is bumped

**Comment by davidchambers** [comment]
@zhangchiqing, see https://github.com/sanctuary-js/sanctuary-set/pull/1#issuecomment-258782119 if you'd like to know how and when the readme is updated. :)

**Pull request merged**